File: benchmarks/python/todo.py
def show(items: list[str]) -> None:
    if len(items) == 0:
        print("nothing to do!")
        return
    for i, entry in enumerate(items, start=1):
        print(f"{i}. {entry}")


def main() -> None:
    items: list[str] = []
    while True:
        try:
            command = input("todo> ")
        except EOFError:
            return
        if command == "quit":
            print("bye")
            return
        if command == "list":
            show(items)
        elif command.startswith("add "):
            items.append(command.split("add ", 1)[1])
            print("added")
        elif command.startswith("done "):
            piece = command.split("done ", 1)[1]
            try:
                n = int(piece)
            except ValueError:
                print("say which number to finish")
                continue
            try:
                items.pop(n - 1)
                print("done!")
            except IndexError:
                print(f"Cannot remove item {n} - the list has {len(items)} item(s).")
        else:
            print("commands: add <thing>, done <number>, list, quit")


if __name__ == "__main__":
    main()

File: benchmarks/rust/todo.rs
use std::io::{self, Write};

fn ask(prompt: &str) -> Option<String> {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut line = String::new();
    let read = io::stdin().read_line(&mut line).ok()?;
    if read == 0 {
        return None;
    }
    Some(line.trim_end().to_string())
}

fn show(items: &[String]) {
    if items.is_empty() {
        println!("nothing to do!");
        return;
    }
    for (index, entry) in items.iter().enumerate() {
        println!("{}. {}", index + 1, entry);
    }
}

fn main() {
    let mut items: Vec<String> = Vec::new();
    loop {
        let Some(command) = ask("todo> ") else {
            return;
        };
        if command == "quit" {
            println!("bye");
            return;
        } else if command == "list" {
            show(&items);
        } else if command.starts_with("add ") {
            items.push(command.replacen("add ", "", 1));
            println!("added");
        } else if command.starts_with("done ") {
            let piece = command.replacen("done ", "", 1);
            let Ok(n) = piece.parse::<usize>() else {
                println!("say which number to finish");
                continue;
            };
            if n == 0 || n > items.len() {
                println!("Cannot remove item {} - the list has {} item(s).", n, items.len());
            } else {
                items.remove(n - 1);
                println!("done!");
            }
        } else {
            println!("commands: add <thing>, done <number>, list, quit");
        }
    }
}
